### PR TITLE
[VAN-391] Add a pass to transform LocationRule::Mapped -> LocationRule::Indexed

### PR DESCRIPTION
**Pull request by iangneal**
Add a pass to convert `LocationRule::Mapped` to `LocationRule::Indexed`. Works for `LocationRule::Mapped` with `indexes.len() == 1`, which I think is all that is supported in the `BucketInterpreter` as well.

### CODE REVIEW

File: circuit_passes/src/passes/mapped_to_indexed.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                LocationRule::Mapped { .. } => unreachable!()
+            }
+            return indexed_rule.clone();
+        }
+        match location_rule {
+            LocationRule::Indexed { location, template_header } => LocationRule::Indexed {
+                location: self.transform_instruction(location),
+                template_header: template_header.clone(),
+            },
+            LocationRule::Mapped { .. } => unreachable!()
```
Comment (0xddom) [inline circuit_passes/src/passes/mapped_to_indexed.rs:175]: Is this to ensure we never produce a `LocationRule::Mapped`?

**Review by 0xddom — CHANGES_REQUESTED** [review]

File: circom/src/compilation_user.rs [hunk 2]
```diff
@@ -64,6 +64,7 @@ pub fn compile(config: CompilerConfig, program_archive: ProgramArchive, prime: &
         // Only run this passes if we are going to generate LLVM code
         let pm = PassManager::new();
         circuit = pm
+            .schedule_mapped_to_indexed_pass(prime)
             .schedule_loop_unroll_pass(prime)
             .schedule_conditional_flattening_pass(prime)
             .schedule_simplification_pass(prime)
```
[the remark this answers is not in the document]
Comment (iangneal) [reply circom/src/compilation_user.rs]: Can do!

File: circuit_passes/src/passes/mapped_to_indexed.rs (hunk 1 above)
Comment (0xddom) [reply circuit_passes/src/passes/mapped_to_indexed.rs:175]: Perfect, good call

**Review by 0xddom — APPROVED** [review]
Perfect. Please move this from draft and we can merge it

**Pull request merged**